src/avram/polymorphic.cr

```diff
@@ -0,0 +1,144 @@
+  #
+  # > *Important note:* the `belongs_to` associations must be nilable since only
+  # > one polymorphic can be set at a time. Remember to also make the association
+  # > nilable in the migration, for example: `add_belongs_to photo : Photo?`
+  #
+  # ```
+  # class Comment < BaseModel
+  #   table do
+  #     belongs_to photo : Photo?
+  #     belongs_to video : Video?
+  #     polymorphic :commentable, associations: [:photo, :video]
+  #   end
+  # end
+  #
+  # class Photo < BaseModel
+  #   table do
+  #     has_many comments : Comment
+  #   end
+  # end
+  #
+  # class Photo < BaseModel
+  #   table do
+  #     has_many comments : Comment
+  #   end
+  # end
+  # ```
+  #
+  # `has_many` calls work just like normal and so do `belongs_to`.
+  #
+  # `polymorphic` just set up methods for validation, preloading, and getting
+  # the polymorphic association more easily.
+  #
+  # ## Validations
+  #
+  # `polymorphic` will add a validation so that exactly one of the associations
+  # is present. In this case either `photo_id` or `video_id`, but not both.
+  #
+  # If `optional` is set to true, it will allow the associations to be `nil`, but
+  # will still not allow more than one association to be set at a time.
+  #
+  # ## Preloading
+  #
+  # `polymorphic` defines a `preload_{polymorphic_name}` on the query.
+  #
+  # In this example it would define `preload_commentable` on `Comment::BaseQuery`.
+  # Under the hood this is preloading each belongs_to. In this example
+  # `Comment::BaseQuery.preload_commentable` is like calling
+  # `Comment::BaseQuery.preload_photo.preload_video`.
+  #
+  # ## Model getter
+  #
+  # Adds `{polymorphic_name}` and `{polymorphic_name}!` getters to the model.
+  #
+  # In this example `commentable` and `commentable!`. Calling these will return
+  # either the `video` or the `photo`. The `!` version will allow lazy loading.
+  # The non `!` version requires preloading with `preload_{polymorphic_name}.
+  #
+  # If `optional` is set to true these methods will return `nil` if nothing
+  # is associated.
+  #
+  # ```
+  # Comment::BaseQuery.new.each do |comment|
+  #   comment.commentable! # Returns the photo or video
+  # end
+  # ```
+  #
+  # If you know which association you want you can just call it directly like
+  # you would with any other `belongs_to`:
+  #
+  # ```
+  # photo.comments.each do |comment|
+  #   # We know it is for a photo, so we can call it directly
+  #   # Note that Crystal will think this might be `nil` since the `belongs_to`
+  #   # is nilable. So you'll need to call `not_nil!` or handle the `nil` case
+  #   # with an `if`
+  #   comment.photo!.not_niL!
+  # end
+  # ```
+  #
+  # # Saving/updating polymorphic associations
+  #
+  # Polymorphic associations are saved just like your normally would with a
+  # `belongs_to` association:
+  #
+  # ```
+  # photo = PhotoQuery.first
+  # # Save a comment for this photo
+  # Comment::SaveOperation.create!(photo_id: photo.id)
```

Comment:
The setting of the polymorphic model surprised me when I saw it in the tests, so it might be worth showing the example of the other model being set.

Reply:
Can you clarify what surprised you and what you’d like to see to make this clearer? I’m not sure I understood

Reply:
Ah sorry. I think I wouldn't have assumed that you could associate a Comment with a Photo by setting its ID when it's polymorphic. I think I would have assumed `Comment::SaveOperation.create!(commentable: photo)`. Because you can set either the photo id or the video_id also, it'd love to see another line below:

```crystal
Comment::SaveOperation.create!(video_id: video.id)
```

Reply:
Oh I see. So make it clear that’s how you associate them by showing both associations being set, right?

Reply:
Yup!